blackboxopt/optimizers/botorch_base.py

```diff
@@ -166,27 +166,48 @@ def to_numerical(
     return X, Y
 
 
-def init_af_opt_kwargs(af_opt_kwargs: Optional[dict]) -> dict:
-    """Provide default initialization for the acquisition function optimizer
-    configuration. Ensure that all mandatory fields are set to be used by BoTorch.
+def _init_acquisition_function_optimizer(
+    search_space: ps.ParameterSpace,
+    af_opt_kwargs: Optional[dict],
+    torch_dtype: torch.dtype,
+) -> Callable[[AcquisitionFunction], Tuple[torch.Tensor, torch.Tensor]]:
+    """Prepare either BoTorch's `optimize_acqf_discrete` or `optimize_acqf` depending
+    on whether the search space is fully discrete or not and set required defaults if
+    not overridden by `af_opt_kwargs`.
 
     Args:
-        af_opt_kwargs: Acquisition function configuration.
+        search_space: Search space used for optimization.
+        af_opt_kwargs: Acquisition function optimizer configuration.
+        torch_dtype: Torch tensor type.
 
     Returns:
-        Acquisition function optimizer configuration, applicable for BoTorch's
-            optimizer.
+        Acquisition function optimizer that takes an acquisition function and returns a
+        candidate with its associate acquisition function value.
     """
+    kwargs = {} if af_opt_kwargs is None else af_opt_kwargs.copy()
 
-    af_opt_config = {} if af_opt_kwargs is None else af_opt_kwargs
-
-    # number of initial samples during AF optimization
-    af_opt_config.setdefault("raw_samples", 1024)
-
-    # number of restarts during AF optimization
-    af_opt_config.setdefault("num_restarts", 4)
-
-    return af_opt_config
+    is_fully_discrete_space = not any(
+        search_space[n]["parameter"].is_continuous
+        for n in search_space.get_parameter_names()
+    )
+    if is_fully_discrete_space:
+        choices = torch.Tensor(
+            [
+                search_space.to_numerical(search_space.sample())
+                for _ in range(kwargs.pop("n_samples", 5_000))
```

Comment:
This also bothered me about the existing arguments and I'm happy to not follow the status quo but change something about it. The easiest approach would be to add it to the `__init__` docstring, but this is also prone to getting out of sync. On the other hand, adding explicit arguments also seems unreasonable. My preference would be adding it to the optimizer `__init__` docstring, how do you feel about that?

Reply:
There's also the option to use a [`TypedDict`](https://docs.python.org/3/library/typing.html#typing.TypedDict), which could expose some of the information, but I have no idea how this will be rendered in docs:

```python
class AfOptKwargs(TypedDict):
    """Optional Acquisition Function Arguments

    Attributes:
        n_samples: Number of samples drawn from grid...
    """
    n_samples: int
    ...

def _acquisition_function_optimizer_factory(
    search_space: ps.ParameterSpace,
    af_opt_kwargs: Optional[AfOptKwargs],
):
    ...
```

Reply:
IIRC TypedDict's don't allow more than the specified keys to be set, though, is that right? And here we specifically rely on expert users being able to put in kwargs for the respective underlying torch af optimizer.

Reply:
Yeah, I think you are right, so it's probably not the right fit...

Reply:
I elaborated a bit more in the docstring, and would move on with the PR.